contracts/Portfolio.sol

```diff
@@ -828,67 +910,95 @@ abstract contract PortfolioVirtual is Objective {
-                if (debited > 0) {
-                    emit DecreaseUserBalance(msg.sender, token, debited);
-                    emit DecreaseReserveBalance(token, debited);
-                }
-
-                // Outstanding amount must be transferred in.
-                if (remainder > 0) {
-                    _payments.push(
-                        Payment({
-                            token: token,
-                            amount: remainder,
-                            balance: Account.__balanceOf__(token, address(this))
-                        })
-                    );
-                }
+            // Reserves were increased, we paid a debit, therefore need to decrease reserves by `debited` amount.
+            if (debited > 0) {
+                emit DecreaseUserBalance(msg.sender, token, debited);
+                emit DecreaseReserveBalance(token, debited);
             }
 
+            // Only `credited` or `remainder` can be non-zero.
+            // Outstanding amount must be transferred in to `address(this)`.
+            // Untracked credits must be transferred out to `msg.sender`.
+            _payments.push(
+                Payment({
+                    token: token,
+                    amountTransferTo: credited, // Reserves are not tracking some tokens.
+                    amountTransferFrom: remainder, // Reserves need more tokens.
+                    balance: Account.__balanceOf__(token, address(this))
+                })
+            );
+
             // Token considered fully accounted for.
             __account__.warm.pop();
             unchecked {
                 --i; // Cannot underflow because loop exits at 0!
             }
         } while (i != 0);
 
-        // Uses `token.transferFrom(msg.sender, amount)` to pay for the outstanding debits.
+        // Uses `token.transferFrom(msg.sender, address(this), amount)` to pay for the outstanding debits.
+        // Uses `token.transfer(msg.sender, amount)` to pay out the untracked credits.
         Payment[] memory payments = _payments;
         uint256 px = payments.length;
         while (px != 0) {
             uint256 index = px - 1;
-            Account.__dangerousTransferFrom__(
-                payments[index].token, address(this), payments[index].amount
+            address token = payments[index].token;
+            uint8 decimals = IERC20(token).decimals();
+
+            (uint256 amountTransferTo, uint256 amountTransferFrom) = (
+                payments[index].amountTransferTo,
+                payments[index].amountTransferFrom
             );
-            unchecked {
-                --px; // Cannot underflow because loop exits at 0!
+
+            // Scale WAD -> Decimals.
+            amountTransferTo = amountTransferTo.scaleFromWadDown(decimals);
+            amountTransferFrom = amountTransferFrom.scaleFromWadDown(decimals);
```

Comment:
would be safer to round up for `amountTransferFrom` but this could lead to unexpected reverts:

swap X -> USDC -> Z. if you don't use `useMax` for the second swap, will end up with a USDC dust amount which is then attempted to be transferred in from the user. Need to think more about if there are other DoS issuess or swap arb issues with not rounding up here.